### DISCUSSION TITLE
customize validator to reject NaN, Infinity

### DISCUSSION THREAD

**Discussion by stealthrabbi** [python-jsonschema/jsonschema#1010] [Q&A] [answered]
I have a Draft7Validator I use to validate some messages.  `decode_schema` calls json.loads() and then `check_schema` on the validator. I then use the validator to validate incoming messages. This works fine, except for handling NaNs and Inifity on number types. I'd like those to be rejected, but I am not sure how to do it with my current setup.

I have seen `validators.extends()`, but I am not sure how to use that method to to set the primary schema and ref schemas. 

```python
primary_schema_object = decode_schema(json_schema_text)
...
self.validator = Draft7Validator(primary_schema_object)
self.validator.resolver.store = reference_schema_objects
```

``` python
    def is_my_int(self, checker, instance):
       # example - change this
        return instance == 4 
```        

When I attempt to use validators.extend() and validate it, I get an error as shown below:

``` python
        type_checker = (
            Draft7Validator.TYPE_CHECKER.redefine("number", self.is_my_int),
        )
        self.validator = validators.extend(Draft7Validator, type_checker=type_checker)
        self.validator(primary_schema_object)
```

```
    def validate(self, *args, **kwargs):
>       for error in self.iter_errors(*args, **kwargs):
E       AttributeError: 'dict' object has no attribute 'iter_errors'
```

**Comment by Julian** [comment, marked as answer]
There may be more than one problem, but there's certainly at least one simple one, which is you did:

```
type_checker = (
            Draft7Validator.TYPE_CHECKER.redefine("number", self.is_my_int),
        )
```

probably using the parentheses to break up the line, but that's a 1-element tuple, because you have a comma there.

So you've passed a tuple in as a type checker, when the right type for that argument is an instance of `TypeChecker` (which `redefine` returns).

You want just `type_checker = Draft7Validator.TYPE_CHECKER.redefine("number", self.is_my_int)` or if you want it on multiple lines remove the comma.

Beyond that it'd be helpful to see a bit more of a fully functioning example to offer advice, because yeah you're in the right spot, if you want to reject those you should redefine `number` (and/or `integer`) to return `False` for `math.isnan` and `math.isinf`, and then use `ModifiedValidator = extend(Draft7Validator, type_checker=Draft7Validator.TYPE_CHECKER.redefine("number", self.is_my_int)); self.validator = ModifiedValidator(schema, store=reference_schema_objects)` rather than using `Draft7Validator` as-is.

**Reply by stealthrabbi** [reply]
Thanks, the piece I was missing (figured out right before you posted) was that the return type of `extend` is returning a class, not an instance, so I had to init it as you've done. I ended up with the following. And yes, I missed the `,` at the end making it a tuple.


``` python
        type_checker = Draft7Validator.TYPE_CHECKER.redefine(
            "number", self._is_finite_number
        )
        extended = validators.extend(Draft7Validator, type_checker=type_checker)
        self.validator = extended(primary_schema_object)
        self.validator.resolver.store = reference_schema_objects
```        


``` python
    def _is_finite_number(self, checker, instance):
        return Draft7Validator.TYPE_CHECKER.is_type(
            instance, "number"
        ) and math.isfinite(instance)
```
